processing/src/main/java/org/apache/druid/segment/writeout/FileWriteOutBytes.java

```diff
@@ -33,22 +33,25 @@
 import java.nio.channels.FileChannel;
 import java.nio.channels.WritableByteChannel;
 
-final class FileWriteOutBytes extends WriteOutBytes
+public final class FileWriteOutBytes extends WriteOutBytes
 {
   private final File file;
   private final FileChannel ch;
   private long writeOutBytes;
 
-  /** Purposely big-endian, for {@link #writeInt(int)} implementation */
-  private final ByteBuffer buffer = ByteBuffer.allocate(4096); // 4K page sized buffer
+  /**
+   * Purposely big-endian, for {@link #writeInt(int)} implementation.
+   * Direct because there is a material difference in performance when writing direct buffers
+   */
+  private final ByteBuffer buffer = ByteBuffer.allocateDirect(32768); // 32K page sized buffer
```

Comment:
this will leak direct memory, it needs to be grabbed from a pool or like use `ByteBufferUtils.free` https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/java/util/common/ByteBufferUtils.java#L199 when we are done with it.

We should almost never be calling `ByteBuffer.allocateDirect` in code unless tracking it

Reply:
also, i'd be interested in seeing the measurements between direct and heap memory here, afaik direct buffers are _slower_ when accessed from inside the JVM, though since we are flushing it to a file here maybe it makes up for it?

Reply:
Are you sure about the memory leak?  `DirectByteBuffer` creates a `Cleaner` which watches the references and frees the memory when it is GC'd.  There shouldn't be a leak from creating a thing here.  It is true that the memory will be used for some time before GC is invoked, but this shouldn't create a leak.

In terms of performance, the updates here were done in response to profiling data.  Part of the speed up comes from the larger page size, another part from the direct flushing.

Reply:
i guess what i'm worried about is like the nested column serializer, which makes child writeout mediums for the nested columns and closes them after the nested column is finished to prevent an explosion of temp files. I'm hoping there wont be an explosion of direct buffers that are dependent on GC to clean up.. it might be cool, but also might not be?

Reply:
How about we modify `FileWriteoutBytes` to accept a `Closer` which `TmpFileSegmentWriteoutMedium` can pass its `Closer` in and `FileWriteoutBytes` can register calling the cleaner this direct buffer with the `Closer`, similar to what `OffHeapMemorySegmentWriteOutMedium` is doing with its `DirectByteBufferWriteOutBytes`.

Reply:
Made this change.